oslo/torch/nn/parallel/data_parallel/zero/sharded_optim/_utils.py

```diff
@@ -0,0 +1,270 @@
+
+    Args:
+        flat (torch.Tensor): The flattened tensor.
+        tensors (Iterable[torch.Tensor]): The original tensor shape.
+
+    Returns:
+        torch.Tensor: The unflattened tensor.
+    """
+    return _unflatten_dense_tensors(flat, tensors)
+
+
+def calculate_global_norm_from_list(norm_list: List[float]) -> float:
+    """
+    Compute the total global norm from a list of norms.
+    Args:
+        norm_list (List[float]): List of norms to compute the total global norm from.
+    Returns:
+        float: Total global norm.
+    """
+    total_norm = 0.0
+    for norm in norm_list:
+        total_norm += norm**2.0
+    return math.sqrt(total_norm)
+
+
+def reduce_tensor_dp_group(
+    tensor: torch.Tensor,
+    dtype: Optional[torch.dtype] = None,
+    dst_local_rank: Optional[int] = None,
+    dst_global_rank: Optional[int] = None,
+    group: Optional[dist.ProcessGroup] = None,
+) -> torch.Tensor:
+    """
+    Reduce a tensor in the data parallel process group.
+
+    Args:
+        tensor (torch.Tensor): Tensor to reduce/all-reduce.
+        dtype (Optional[torch.dtype]): Data type used in communication.
+        dst_local_rank (Optional[int]): Local rank of the destination node.
+        dst_global_rank (Optional[int]): Global rank of the destination node.
+        group (Optional[dist.ProcessGroup]): Process group for the reduction.
+
+    Returns:
+        tensor (torch.Tensor): The reduced tensor.
+    """
+
+    # use the original dtype
+    if dtype is None:
+        dtype = tensor.dtype
+
+    # cast the data to specified dtype for reduce/all-reduce
+    if tensor.dtype != dtype:
+        tensor_to_reduce = tensor.to(dtype)
+    else:
+        tensor_to_reduce = tensor
+
+    world_size = dist.get_world_size(group=group)
+    tensor_to_reduce.div_(world_size)
+
+    # if rank is None, all reduce will be used
+    # else, reduce is used
+    use_all_reduce = dst_local_rank is None
+
+    if use_all_reduce:
+        dist.all_reduce(tensor_to_reduce, group=group)
+    else:
+        dist.reduce(tensor=tensor_to_reduce, dst=dst_global_rank, group=group)
+
+    # recover the original dtype
+    if tensor.dtype != dtype and tensor is not tensor_to_reduce:
+        local_rank = dist.get_rank(group=group)
+        if use_all_reduce or dst_local_rank == local_rank:
+            tensor.copy_(tensor_to_reduce)
+
+    return tensor
+
+
+def has_inf_or_nan(tensor: torch.Tensor) -> bool:
+    """
+    Check if a tensor has any NaN or Inf values.
+
+    Args:
+        tensor (torch.Tensor): Tensor to check.
+
+    Returns:
+        bool: True if the tensor has NaN or Inf values, False otherwise.
+    """
+    try:
+        fp32_tensor = tensor.float()
+    except RuntimeError as exception:
+        if "value cannot be converted" not in exception.args[0]:
+            raise exception
+        return True
+    else:
+        if torch.isinf(fp32_tensor).any() or torch.isnan(fp32_tensor).any():
+            return True
```

Comment:
```suggestion
        if torch.isfinite(fp32_tensor).any():
            return True
```

Reply:
```suggestion
        if torch.isinf(fp32_tensor).any() or torch.isnan(fp32_tensor).all():
            return False
        return True
```

Reply:
I think this change is trivial, haha. 😂 Never mind this.